### PR TITLE
Add FileService MinIO documentation

### PR DESCRIPTION
**Pull request by Thymu**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/systemlink-operations-handbook/blob/master/CONTRIBUTING.md).

Adding the documentation on using MinIO as a storage provider for the SystemLink FileService.

The official documentation covers how to use Amazon S3 as a cloud storage provider, which is what our implementation primarily addresses. Beyond that, MinIO can also be used with a few minor tweaks. This, however, is not part of the documention on ni.com.

Here's a preview: https://github.com/ni/systemlink-operations-handbook/blob/69cbc455c0b57f6e57b9b728af01d22fb677d4c5/handbook/FileService-MinIO/FileService-MinIO.md

The "note"-block is not shown correctly in that preview, it'll look like this:
image

### CODE REVIEW

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:9]: Consider including numbered steps in this section and the next. That way, it's easier for people to follow.

I'd also check with @GrKuehnH to avoid content duplication. I think we're fine for this one, but just as a best practice : )

Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:9]: I think numbered steps for the first section is fine. I'm hesitant to do so for the second section, since this is just one step after all. If numbered steps is preferred, I'd rather re-write the section. Would you like me to provide an alternative phrasing for this section that is numbered? @jjaco2

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:33]: Maybe we could ask Heidi to include these settings as an optional step in the linked S3 documentation. That would make this section more straightforward and truly one step.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [reply handbook/FileService-MinIO/FileService-MinIO.md:9]: This looks good! Adding a few tweaks here and there, just for NI style and stuff. For the second section, I agree with you looking at it again. Below I've suggested a change that could make the second section more direct while also not having to number anything -- so minimal rewriting on your part!

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:11]: For minimalism, can take out 'In order to run the MinIO server,'

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:17]: 'server, use a web browser to connect to the endpoint that the command line prints out. 

(possible step 3?) 3.  Create a bucket using the '+' button...'

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:21]: This note seems more important than the other 2 -- like it might prevent a user from completing the task successfully if they do not apply a bucket region. If this is the case, configuring this setting should be its own step.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:24]: For detailed information

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:28]: instructions

Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:28]: For consistency, use File Service as 2 words here and throughout

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:11]: That raises the question if I can also modify the two sentences just above of that line which both start in "To use MinIO". I feel like my English is limited to just one way of starting a sentence :D 

Any suggestions? I think at least one of the sentence should read differently.

Comment (jjaco2) [reply handbook/FileService-MinIO/FileService-MinIO.md:11]: I left it in the other instances, since it's not a step but more of an overview of what the user generally needs to do. So I think you're fine!

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (prestwick) [inline handbook/FileService-MinIO/FileService-MinIO.md:30]: I don't think we need to state they are undocumented since we documenting them here :-). Perhaps use "additional settings" rather than "undocumented settings"

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (prestwick) [inline handbook/FileService-MinIO/FileService-MinIO.md:32]: Consider documenting this as a workflow with an order list rather than unordered bullets.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (prestwick) [inline handbook/FileService-MinIO/FileService-MinIO.md:37]: This code snippet is so short that I don't we need to have it in a collapsable. Consider using a code fence instead.

**Review by prestwick — CHANGES_REQUESTED** [review]
I would like to see this document expanded such that by the end of the it the reader knows how to leverage MinIO for GCP and Azure file storage services.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [reply handbook/FileService-MinIO/FileService-MinIO.md:26]: Addressed in my review ^-^

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [reply handbook/FileService-MinIO/FileService-MinIO.md:32]: +1, but only if we can't roll this into the SPD content linked above.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:33]: I talked to Stefan Romainczyk and we don't want to mention MinIO in our official documentation.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md]: The link to the documentation is a couple lines further down. I will see to moving that further to the top.

**Comment by Thymu** [comment]
I've made a few changes to this review, addressing a couple of points mentioned by Mark. However, I dismissed a handful of requests that ask for a major change concerning GCS and other cloud storages. We will only address minio for the moment and update in future if required. This course of action was approved by Stefan Romainczyk.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (JD-Robertson) [inline handbook/FileService-MinIO/FileService-MinIO.md:29]: For deployment, I assume the user will want a .bat file or powershell script to launch the server with the appropriate configuration. Should we provide example source for such a file here, allowing the user to easily copy-paste what they need to modify and run on their system?

**Review by JD-Robertson — APPROVED** [review]

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (cameronwaterman) [inline handbook/FileService-MinIO/FileService-MinIO.md:7]:
```suggestion
To use MinIO as a storage provider, set up the MinIO server on a system that you would like to upload files to. This can be the same machine as the SystemLink server or your own dedicated server. Then, configure the File Service to use that server.
```

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (cameronwaterman) [inline handbook/FileService-MinIO/FileService-MinIO.md:24]:
```suggestion
3. When starting the MinIO server for the first time, use a web browser and navigate to the URL that the command line prints out. Create a bucket by clicking the `+` button in the right bottom corner.
```

The terminology used in this context might be a bit too in the weeds for a IT admin wanting to set this up. I suggest we put things more in layman's terms.

After the point about navigating in a browser, maybe we should add a note about what the user should expect to see at this point?

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (cameronwaterman) [inline handbook/FileService-MinIO/FileService-MinIO.md:32]: Does `us-east-1` apply for all global customers? I know you said it's not necessary but maybe we could provide a bit more detail here.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (cameronwaterman) [inline handbook/FileService-MinIO/FileService-MinIO.md:59]: Do we have an expectation that the user knows how to do this at this point?

**Review by cameronwaterman — APPROVED** [review]

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (JD-Robertson) [reply handbook/FileService-MinIO/FileService-MinIO.md:32]: It sounds like you just need a string here that matches the SL configuration. We should probably make that clear, and that we are using 'us-east-1' in examples for convenience.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:29]: Are you asking for a script file that adds the two required settings "S3BackEndServiceUrl" and "S3ForcePathStyle" to the FileIngestion json file or what is it you're asking for?

At any rate, an example for the json content is already given at the very bottom which they can copy to the file.

Comment (JD-Robertson) [reply handbook/FileService-MinIO/FileService-MinIO.md:29]: No, I am suggesting a script for running the service with the correct environment. Or to just gather the entire command line sequence in one clear code block, as is common practice for these sort of instructions.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md]: >After the point about navigating in a browser, maybe we should add a note about what the user should expect to see at this point?
The sentence would be along the line:
>Now you'll see the bucket in the web UI.

There isn't anything spectacular happening, is this of any help to the user?

Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md]: Thanks for suggesting better phrasing, btw. :)

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:32]: 'us-east-1' is just an arbitrary but valid bucket region. It will work for any customer because minio is not hosted on AWS. I added this reasoning to clarify. I also added a second example just to decouple a bit more from the expectation that it had to be us-east-1.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md]: I do think so.
For the 2020 releases, this information was also part of the official documentation. Since we're about to remove it from the documentation (we improved the customer experience and the phrasing has changed, thus), I just copied it over to this documentation here anyways.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:29]: Okay, just amended the two code lines to a full example.
image

Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:29]: I think that writing a batch file would be overkill, since you had to pass in three arguments, make sure that there are three arguments, make sure that the path is valid, print out a meaningful help.... That's a lot for three simple lines of code.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:13]: Maybe 'to start the server' ?

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:27]: 'NI recommends setting a non-default access key and secret key. To do this, define the 'MINIO_ACCESS_KEY' and 'MINIO_SECRET_KEY' variables.'

Since this is recommended, maybe we should make it an actual step? We could mark it as optional if you want.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:33]: Is this a selection they make in the bucket creation prompt? If it is, this could maybe be another step instead of a note. Could be like:

4. Apply a valid bucket region for the File Service to work correctly. Specify any valid Amazon S3 region...

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:47]: actual

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:42]: Add these MinIO-specific settings to the JSON configuration file at `C:\ProgramData\National Instruments\Skyline\Config\FileIngestion.json`:

**Review by jjaco2 — APPROVED** [review]
Looks great! Just some small comments but really like the new steps.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:27]: Actually, it is *not* NI who recommends this, but minIO themselves. I'd rather keep it as "It is recommended" but can certainly make it a separate step.

Comment (jjaco2) [reply handbook/FileService-MinIO/FileService-MinIO.md:27]: Ah! That makes sense. In that case, we might want to say that minIO recommends it, just to avoid liability on our end! We could even link to somewhere in the minIO help that supports this.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:42]: The settings aren't MinIO-specific. They are just required when using MinIO.
I'd go with "Add the two additional settings to the JSON configuration file...."

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:27]: Moved it upwards in the document, it's step 2 now.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:33]: Actually, this note wasn't correct there any more. Initially the note block was at the bottom of the file. I moved this out to the second paragraph, and introduced sub-enumeration. It's 1b now.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (maxgax) [inline handbook/FileService-MinIO/FileService-MinIO.md:3]: I think this section sounds a bit strange. With the use of "while" I was expecting a contrasting statement to follow. Ex. While we use A, we don't support B. I was thinking this might sound better.
'Since NI officially supports Amazon S3 cloud storage, the File Service will also work with MinIO, a server-side software storage stack that is compatible with Amazon S3.'
You may be able to leave the first part of that sentence off entirely. 
'The File Service will also work with MinIO, a server-side software storage stack that is compatible with Amazon S3.'

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (maxgax) [inline handbook/FileService-MinIO/FileService-MinIO.md:5]: @prestwick - Have we decided on a standard way to call out supported versions for documentation in the operations handbook?

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (maxgax) [inline handbook/FileService-MinIO/FileService-MinIO.md:62]: Shouldn't 'NI SystemLink Server Configuration' be capitalized?

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [reply handbook/FileService-MinIO/FileService-MinIO.md:3]: The current wording makes sense to me since S3 and MinIO are 2 different things, and I think the "while" statement is describing a more back-end process. But your second suggestion would work just fine here, I think.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:13]: > you should change the access key and secret key. You can do that by setting the environment variables `MINIO_ACCESS_KEY` and `MINIO_SECRET_KEY`.

....change the access key and secret key. To do this, define the environment variables...

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:30]: I've never seen the phrase "retain a server." I don't see it much online, either -- does retain mean to keep it running? Or to save it, or something else?

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [inline handbook/FileService-MinIO/FileService-MinIO.md:40]: Leading with the user action here will still make this clearer. Does the user actually need to know that there's _technically_ no bucket region? Maybe something like: 

Apply a valid bucket region. You can use any valid Amazon S3 region...

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (jjaco2) [reply handbook/FileService-MinIO/FileService-MinIO.md:62]: It depends on whether we're talking about the actual thing people launch vs. the abstract concept -- and here, it's the former, so I agree we should capitalize.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (JD-Robertson) [reply handbook/FileService-MinIO/FileService-MinIO.md:3]: I don't believe that describing a background process was the intent, though. I would think "because" if you want to keep the opening phrase, but I also prefer the more concise second option.

**Review by prestwick — CHANGES_REQUESTED** [review]
Looking really good. Unless there is some urgency to pushing this document out, I'd like to hold off publishing this doc until we've published #32. That PR creates a hierarchy around data storage as a general topic where mongoDB, file storage, (and maybe) DataFinder index are covered as chapters within that topic. This chapter would fit in as a peer to those documents. If we want to get this out prior to my PR I recommend looking at the folder structure described in #32. Doing this ensures URLs don't change when #32 goes in.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:3]: I dropped the "while" part and used the concise version that Max suggested.

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:30]: Meant "Keep it running". I fixed it

File: handbook/FileService-MinIO/FileService-MinIO.md
Comment (Thymu) [reply handbook/FileService-MinIO/FileService-MinIO.md:9]: I tried using the MinIO Gateway tool to connect to GCS. Unfortunately, the Upload method that we use is not implemented. Consequently, usage is limited to the MinIO server tool. GCS, Azure etc. will not work.

I added a statement on that.

**Review by prestwick — CHANGES_REQUESTED** [review]
> I've made a few changes to this review, addressing a couple of points mentioned by Mark. However, I dismissed a handful of requests that ask for a major change concerning GCS and other cloud storages. We will only address minio for the moment and update in future if required. This course of action was approved by Stefan Romainczyk.

Hi @Thymu, given the constraints that keep us from supporting Azure blob or GCP I suggest we change the name of this document. Something like, **Leverage File Service S3 for On-premises storage**. 

I'd also like this document to be moved into the `data-stores` directory rather than kept in its current directory. 

Please make these couple of changes, and I will approve the PR.

**Comment by Thymu** [comment]
> > I've made a few changes to this review, addressing a couple of points mentioned by Mark. However, I dismissed a handful of requests that ask for a major change concerning GCS and other cloud storages. We will only address minio for the moment and update in future if required. This course of action was approved by Stefan Romainczyk.
> 
> Hi @Thymu, given the constraints that keep us from supporting Azure blob or GCP I suggest we change the name of this document. Something like, **Leverage File Service S3 for On-premises storage**.
> 
> I'd also like this document to be moved into the `data-stores` directory rather than kept in its current directory.
> 
> Please make these couple of changes, and I will approve the PR.



Done :)

**Review by usteinha — APPROVED** [review]

**Review by prestwick — APPROVED** [review]

**Pull request merged**
